File: sdks/java/src/main/java/com/blok/blok/nodes/BlokErrorDemoNode.java
package com.blok.blok.nodes;

import com.blok.blok.errors.BlokError;
import com.blok.blok.errors.BuildContextSnapshot;
import com.blok.blok.node.NodeHandler;
import com.blok.blok.types.Context;

import java.io.IOException;
import java.time.Duration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Example node demonstrating the structured {@link BlokError} API per master
 * plan §17.
 *
 * <p>Used by the cross-language E2E test
 * ({@code core/runner/__tests__/integration/runtimes/java-grpc.integration.test.ts})
 * to verify that a Java-side structured error flows through the gRPC wire
 * to the runner with every field preserved (category, severity, code,
 * remediation, retryable hints, cause chain, context snapshot).
 *
 * <p>Triggered via the {@code mode} config:
 * <ul>
 *   <li>{@code mode="dependency"} (default) — throws
 *       {@code BlokError.dependency()} with a cause chain rooted in an
 *       {@link IOException}.</li>
 *   <li>{@code mode="rate-limit"} — throws {@code BlokError.rateLimit()}
 *       with {@code retryAfterMs}.</li>
 *   <li>{@code mode="validation"} — throws {@code BlokError.validation()}
 *       with {@code details.issues} (Zod-style).</li>
 *   <li>{@code mode="ok"} — returns success.</li>
 * </ul>
 */
public class BlokErrorDemoNode implements NodeHandler {

    @Override
    public Object execute(Context ctx, Map<String, Object> config) {
        String mode = "dependency";
        if (config != null) {
            Object m = config.get("mode");
            if (m instanceof String s && !s.isEmpty()) mode = s;
        }

        if ("ok".equals(mode)) {
            Map<String, Object> ok = new HashMap<>();
            ok.put("ok", true);
            ok.put("language", "java");
            return ok;
        }

        Map<String, Object> snapshot = BuildContextSnapshot.of(
                config != null ? config : Map.of(),
                ctx != null && ctx.getVars() != null ? ctx.getVars() : Map.of()
        );

        if ("rate-limit".equals(mode)) {
            Map<String, Object> details = new HashMap<>();
            details.put("limit", 5000);
            details.put("remaining", 0);
            throw BlokError.rateLimit()
                    .code("UPSTREAM_RATE_LIMITED")
                    .message("Upstream API returned 429")
                    .description("GitHub API rate limit hit (5000 req/hr).")
                    .remediation("Wait until the X-RateLimit-Reset header timestamp.")
                    .retryAfterMs(60_000)
                    .docUrl("https://docs.example.com/errors/rate-limit")
                    .details(details)
                    .contextSnapshot(snapshot)
                    .build();
        }

        if ("validation".equals(mode)) {
            Map<String, Object> details = new HashMap<>();
            details.put("issues", List.of(
                    Map.of("path", List.of("email"), "message", "Required"),
                    Map.of("path", List.of("name"), "message", "Required")
            ));
            throw BlokError.validation()
                    .code("VALIDATION_FAILED")
                    .message("2 validation issues")
                    .description("Inputs didn't match the node's schema.")
                    .remediation("Provide both `email` and `name`.")
                    .details(details)
                    .contextSnapshot(snapshot)
                    .build();
        }

        // default: dependency with a cause chain rooted in an IOException.
        IOException cause = new IOException("[Errno 61] Connection refused");
        Map<String, Object> details = new HashMap<>();
        details.put("host", "db.internal");
        details.put("port", 5432);
        details.put("timeout_ms", 5000);
        throw BlokError.dependency()
                .code("POSTGRES_CONNECT_TIMEOUT")
                .message("Could not connect to Postgres within 5s")
                .description("Tried host=db.internal port=5432; timeout=5000ms")
                .remediation("Check DATABASE_URL env var and network reachability")
                .cause(cause)
                .retryable(true)
                .retryAfter(Duration.ofSeconds(5))
                .docUrl("https://docs.example.com/errors/POSTGRES_CONNECT_TIMEOUT")
                .details(details)
                .contextSnapshot(snapshot)
                .build();
    }
}

File: sdks/csharp/src/Blok.Core/Nodes/BlokErrorDemoNode.cs
using System.Text.Json;
using Blok.Core.Errors;
using Blok.Core.Node;

namespace Blok.Core.Nodes;

/// <summary>
/// Example node demonstrating the structured <see cref="BlokError"/> API per
/// master plan §17.
///
/// <para>Used by the cross-language E2E test
/// (<c>core/runner/__tests__/integration/runtimes/csharp-grpc.integration.test.ts</c>)
/// to verify that a C#-side structured error flows through the gRPC wire to
/// the runner with every field preserved (category, severity, code,
/// remediation, retryable hints, cause chain, context snapshot).</para>
///
/// <para>Triggered via the <c>mode</c> config:</para>
/// <list type="bullet">
///   <item><c>mode="dependency"</c> (default) — throws <c>BlokError.Dependency()</c>
///     with a cause chain rooted in an <see cref="IOException"/>.</item>
///   <item><c>mode="rate-limit"</c> — throws <c>BlokError.RateLimit()</c>
///     with <c>retry_after_ms</c>.</item>
///   <item><c>mode="validation"</c> — throws <c>BlokError.Validation()</c>
///     with <c>details.issues</c>.</item>
///   <item><c>mode="ok"</c> — returns success.</item>
/// </list>
/// </summary>
public class BlokErrorDemoNode : INodeHandler
{
    public Task<JsonElement> ExecuteAsync(Types.Context ctx, Dictionary<string, JsonElement> config)
    {
        var mode = "dependency";
        if (config.TryGetValue("mode", out var modeVal) && modeVal.ValueKind == JsonValueKind.String)
        {
            mode = modeVal.GetString() ?? "dependency";
        }

        if (mode == "ok")
        {
            var okJson = JsonSerializer.Serialize(new { ok = true, language = "csharp" });
            return Task.FromResult(JsonDocument.Parse(okJson).RootElement.Clone());
        }

        // Build the snapshot from the resolved inputs + ctx.vars.
        var configMap = ConfigToObjectMap(config);
        var snapshot = BuildContextSnapshot.Of(configMap, ctx.Vars ?? new Dictionary<string, object?>());

        if (mode == "rate-limit")
        {
            throw BlokError.RateLimit()
                .Code("UPSTREAM_RATE_LIMITED")
                .Message("Upstream API returned 429")
                .Description("GitHub API rate limit hit (5000 req/hr).")
                .Remediation("Wait until the X-RateLimit-Reset header timestamp.")
                .RetryAfterMs(60_000)
                .DocUrl("https://docs.example.com/errors/rate-limit")
                .Details(new Dictionary<string, object?>
                {
                    ["limit"] = 5000,
                    ["remaining"] = 0,
                })
                .ContextSnapshot(snapshot)
                .Build();
        }

        if (mode == "validation")
        {
            throw BlokError.Validation()
                .Code("VALIDATION_FAILED")
                .Message("2 validation issues")
                .Description("Inputs didn't match the node's schema.")
                .Remediation("Provide both `email` and `name`.")
                .Details(new Dictionary<string, object?>
                {
                    ["issues"] = new List<Dictionary<string, object?>>
                    {
                        new() { ["path"] = new List<string> { "email" }, ["message"] = "Required" },
                        new() { ["path"] = new List<string> { "name" }, ["message"] = "Required" },
                    },
                })
                .ContextSnapshot(snapshot)
                .Build();
        }

        // default: dependency with a cause chain rooted in an IOException.
        var cause = new IOException("[Errno 61] Connection refused");
        throw BlokError.Dependency()
            .Code("POSTGRES_CONNECT_TIMEOUT")
            .Message("Could not connect to Postgres within 5s")
            .Description("Tried host=db.internal port=5432; timeout=5000ms")
            .Remediation("Check DATABASE_URL env var and network reachability")
            .Cause(cause)
            .Retryable(true)
            .RetryAfter(TimeSpan.FromSeconds(5))
            .DocUrl("https://docs.example.com/errors/POSTGRES_CONNECT_TIMEOUT")
            .Details(new Dictionary<string, object?>
            {
                ["host"] = "db.internal",
                ["port"] = 5432,
                ["timeout_ms"] = 5000,
            })
            .ContextSnapshot(snapshot)
            .Build();
    }

    private static Dictionary<string, object?> ConfigToObjectMap(Dictionary<string, JsonElement> config)
    {
        var map = new Dictionary<string, object?>();
        foreach (var kv in config)
        {
            map[kv.Key] = JsonElementToObject(kv.Value);
        }
        return map;
    }

    private static object? JsonElementToObject(JsonElement el) => el.ValueKind switch
    {
        JsonValueKind.String => el.GetString(),
        JsonValueKind.Number => el.TryGetInt64(out var l) ? (object)l : el.GetDouble(),
        JsonValueKind.True => true,
        JsonValueKind.False => false,
        JsonValueKind.Null => null,
        _ => el.GetRawText(),
    };
}
